eth_account/hdaccount/mnemonic.py

```diff
@@ -0,0 +1,178 @@
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+# WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+#
+
+import binascii
+import hashlib
+import os
+from pathlib import (
+    Path,
+)
+import unicodedata
+
+from eth_utils import (
+    ValidationError,
+    combomethod,
+)
+
+PBKDF2_ROUNDS = 2048
+VALID_SEED_SIZES = [16, 20, 24, 28, 32]
+VALID_WORD_LENGTHS = [12, 15, 18, 21, 24]
+WORDLIST_DIR = Path(__file__).parent / "wordlist"
+
+
+def normalize_string(txt):
+    if isinstance(txt, bytes):
+        utxt = txt.decode("utf8")
+    elif isinstance(txt, str):
+        utxt = txt
+    else:
+        raise ValidationError("String value expected")
+
+    return unicodedata.normalize("NFKD", utxt)
+
+
+class Mnemonic(object):
+    def __init__(self, language):
+        if language not in self.list_languages():
+            raise ValidationError(
+                f'Invalid language choice "{language}", must be one of {self.list_langauges()}'
+            )
+        self.language = language
+        self.radix = 2048
+        with open(WORDLIST_DIR / f"{language}.txt", "r", encoding="utf-8") as f:
+            self.wordlist = [w.strip() for w in f.readlines()]
+        if len(self.wordlist) != self.radix:
+            raise ValidationError(
+                f"Wordlist should contain {self.radix} words, "
+                f"but it contains {len(self.wordlist)} words."
+            )
+
+    @combomethod
+    def list_languages(_Mnemonic):
+        return sorted(Path(f).stem for f in WORDLIST_DIR.rglob("*.txt"))
+
+    @classmethod
+    def detect_language(cls, raw_mnemonic):
+        mnemonic = normalize_string(raw_mnemonic)
+        words = mnemonic.split(" ")
+        languages = cls.list_languages()
+
+        language_counts = {lang: 0 for lang in languages}
+        for lang in languages:
+            wordlist = cls(lang).wordlist
+            for word in words:
+                if word in wordlist:
+                    language_counts[lang] += 1
+
+        # No language had all words match it, so the language can't be fully determined
+        if max(language_counts.values()) < len(words):
+            raise ValidationError("Language not detected")
+
+        # Because certain wordlists share some similar words, we detect the language
+        # by seeing which of the languages have the most hits, and returning that one
+        most_matched_language = max(language_counts.items(), key=lambda c: c[1])[0]
```

Comment:
This happens with Chinese, because simplified can be a subset of traditional. I fixed this, and also the bug with Chinese (it should assume simplified if both are okay) in 6e3ba5d